packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -83,109 +37,178 @@ library AmmJoinProcess
+            deposit.accountID== ctx.accountID &&
+            deposit.tokenID == poolToken.tokenID &&
+            deposit.amount == amount,
+            "INVALID_TX_DATA"
+        );
+
+        ctx.exchange.deposit{value: 0}(
+            address(this),
+            to,
+            poolToken.addr,
+            amount,
+            new bytes(0)
+        );
     }
 
     function _calculateJoinAmounts(
-        AmmData.State    storage S,
         AmmData.Context  memory  ctx,
         AmmData.PoolJoin memory  join
         )
         private
         view
         returns(
-            bool            slippageRequirementMet,
-            uint            poolAmountOut,
+            bool            slippageOK,
+            uint96          mintAmount,
             uint96[] memory amounts
         )
     {
         // Check if we can still use this join
-        amounts = new uint96[](ctx.size);
-        uint _totalSupply = S.totalSupply;
+        amounts = new uint96[](ctx.size - 1);
 
         if (block.timestamp > join.validUntil) {
             return (false, 0, amounts);
         }
 
-        if (_totalSupply == 0) {
-            return(true, ctx.poolTokenInitialSupply, join.maxAmountsIn);
+        if (ctx.totalSupply == 0) {
+            return(true, ctx.poolTokenInitialSupply.toUint96(), join.joinAmounts);
         }
 
-        // Calculate the amount of liquidity tokens that should be minted
-        for (uint i = 0; i < ctx.size; i++) {
+        // Calculate the amount of pool tokens that should be minted
+        bool initialized = false;
+        for (uint i = 1; i < ctx.size; i++) {
             if (ctx.ammExpectedL2Balances[i] > 0) {
-                uint amountOut = uint(join.maxAmountsIn[i])
-                    .mul(_totalSupply) / uint(ctx.ammExpectedL2Balances[i]);
-
-                if (amountOut < poolAmountOut) {
-                    poolAmountOut = amountOut;
+                uint amountOut = uint(join.joinAmounts[i - 1])
+                    .mul(ctx.totalSupply) / uint(ctx.ammExpectedL2Balances[i]);
+
+                if (!initialized) {
+                    initialized = true;
+                    mintAmount = amountOut.toUint96();
+                } else if (amountOut < mintAmount) {
+                    mintAmount = amountOut.toUint96();
                 }
             }
         }
 
-        if (poolAmountOut == 0) {
+        if (mintAmount == 0) {
             return (false, 0, amounts);
         }
 
         // Calculate the amounts to deposit
-        uint ratio = poolAmountOut.mul(ctx.poolTokenBase) / _totalSupply;
+        uint ratio = ctx.poolTokenBase.mul(mintAmount) / ctx.totalSupply;
 
-        for (uint i = 0; i < ctx.size; i++) {
-            amounts[i] = ratio.mul(ctx.ammExpectedL2Balances[i] / ctx.poolTokenBase).toUint96();
+        for (uint i = 1; i < ctx.size; i++) {
+            amounts[i - 1] = ratio.mul(ctx.ammExpectedL2Balances[i] / ctx.poolTokenBase).toUint96();
```

Comment:
Seems like the liquidity token was the first token and this worked differently in previous versions. :)